### DISCUSSION TITLE
Last input radio option gets selected on form reset

### DISCUSSION THREAD

**Discussion by mickaelchanrion** [alpinejs/alpine#3638] [7. Bugs (archived)]
Hello guys 👋
I noticed a behavior I didn't expect using x-model on a radio group.
https://codepen.io/Michkkael/pen/eYQWvXN

```html
<form x-data="{ test: null }">
  <fieldset>
    <legend>Test</legend>
    <div>
      <input x-model="test" type="radio" name="test" value="1" id="test-1" />
      <label for="test-1">Option 1</label>
    </div>
    <div>
      <input x-model="test" type="radio" name="test" value="2" id="test-2" />
      <label for="test-2">Option 2</label>
    </div>
    <div>
      <input x-model="test" type="radio" name="test" value="3" id="test-3" />
      <label for="test-3">Option 3</label>
    </div>
  </fieldset>
  <fieldset>
    Selected: <span x-text="test"></span>
  </fieldset>
  <button type="reset">Reset</button>
</form>
```

Upon form reset event, the last radio option gets selected. IMO the expected result should be the same as the native one: no selected option

**Comment by tvdr** [comment]
Hi, 

as per #3143 the x-model directive adds an eventlistener to the forms reset event. 

`
nextTick(() => el._x_model && el._x_model.set(el.value))
`

On the above example uppon reset, the listener will be invoked for every input, setting the model value to the input value,1, 2, and then 3.
Checkboxes are also affected by this bug.

Please see my [branch](https://github.com/alpinejs/alpine/compare/main...tvdr:alpine:bug/reset-form-radio-and-checkbox) for failing tests demonstrating the bug and a possible solution.

Would happily submit a PR if this would be the correct behaviour.

**Reply by ekwoka** [reply]
This is a bit difficult, because radios and checkboxes may not be null as the default value they reset to.

Seems you'd have to check if they have the actual `checked` property. As far as I'm aware, that property does not reflect actual state, only the initial html state.

**Reply by tvdr** [reply]
You're right, according to the HTML spec , resetting an input works as:
> The [reset algorithm](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-form-reset-control) for [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) elements is to set the [dirty value flag](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-dirty) and [dirty checkedness flag](https://html.spec.whatwg.org/multipage/input.html#concept-input-checked-dirty-flag) back to false, set the [value](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-value) of the element to the [value](https://html.spec.whatwg.org/multipage/input.html#attr-input-value) of the value content attribute, if there is one, or the empty string otherwise, set the [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) of the element to true if the element has a [checked](https://html.spec.whatwg.org/multipage/input.html#attr-input-checked) content attribute and false if it does not, empty the list of [selected files](https://html.spec.whatwg.org/multipage/input.html#concept-input-type-file-selected), and then invoke the [value sanitization algorithm](https://html.spec.whatwg.org/multipage/input.html#value-sanitization-algorithm), if the [type](https://html.spec.whatwg.org/multipage/input.html#attr-input-type) attribute's current state defines one.

So it's using the value and checked content attributes (these are the attributes "visible" on the element, not the IDL properties; el.getAttribute('checked') and el.checked are two different things - not trying to overexplain)

I think no one includes the value or checked attribute for an input when using Alpine (nor does Alpine do it), so when resetting the form, all textinputs will be set to empty string (as per the html spec) and then  `nextTick(() => el._x_model && el._x_model.set(el.value))` will update the model to empty string also.  While this is acceptable, setting radio and checkbox model values to (okay, not null but) empty string doesn't sound so bad either :)

And these are just the inputs,there are also selects, and textareas.

The most perfect solution that i could think of is properly setting the required content attributes  and such on x-model init:

- checked on radios and checkboxes
- value on textinput
- selected on select option(s)
- textnode data for textarea

this way the model value could be upgraded to the current element state to stay in sync.

**Reply by ekwoka** [reply]
> I think no one includes the value or checked attribute for an input when using Alpine

Now that `x-model.fill` exists, this will probably be more common and something that would even be encouraged as the way to handle initial state.

**Comment by ekwoka** [comment]
hmm, I think this should be solved by doing something like with:

#3495

Instead of trying to set the value directly, just triggering a change event on the element and letting the normal behaviors follow it.
